docs/JLBP-21.md

```diff
@@ -0,0 +1,31 @@
+# [JLBP-21] Separate tool classpath from product classpath
+
+The tools we use for Java projects are more often than not themselves
+written in Java. This includes build system such as Maven, javac, and Ant;
+runtime environments such as Tomcat, Hadoop, Beam, and App Engine; and editors
+such as jEdit, Eclipse, and IntelliJ IDEA. Tools like these should
+have their own classpaths from which they load their own code.
+
+It is critical not to mix the classpath of the tool with the classpath of the product.
+Dependencies of the tool should not be dependencies of the product.
+For example, there's no reason `javax.tools.JavaCompiler` should appear in the
+classpath of an MP3 player simply because the product was compiled by `javac`.
+
+Indeed `javac` does not transmit its own dependencies into the products
+it builds, but not all tools are this well behaved. Maven annotation processors
+such as AutoValue and Animal Sniffer are sometimes declared to be dependencies
+of the product itself in the pom.xml. Since they are only needed at
+compile time, they should instead be added to the annotation processor classpath.
```

Comment:
Is it worth linking to https://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#annotationProcessorPaths and https://docs.gradle.org/4.6/release-notes.html#convenient-declaration-of-annotation-processor-dependencies ? I haven't checked whether we normally provide such links to specific tools.

Reply:
added a link to the Maven page.